pdns/dnsrecords.cc

```diff
@@ -489,6 +489,242 @@ string EUI64RecordContent::getZoneRepresentation(bool noDot) const
+        bytes--;
+      } else
+      {
+        // Found non-0 byte, stop trimming
+        break;
+      }
+      v4ip = v4ip >> 8;
+    }
+
+    // Prefix length
+    if (sscanf(subnetstr.c_str(), "%u", &prefix) == 1) {
+      ret->d_prefix=prefix;
+    } else
+          throw MOADNSException("Asked to decode '"+zone+"' as an IPv4 APL record, but found invalid prefix string "+subnetstr);
+    if (prefix > 32)
+      throw MOADNSException("Asked to decode '"+zone+"' as an IPv4 APL record, but found invalid prefix value "+std::to_string(prefix));
+    ret->d_afdlength = bytes;
+
+  } else if (record.find("2:", 0) == 0) { // IPv6
+    unsigned int prefix;
+    int subnet_pos;
+    string ipstr, subnetstr;
+    ComboAddress ca;
+    int bytes;
+
+    ret->d_family = APL_FAMILY_IPV6;
+
+    // Find Netmask
+    subnet_pos = record.rfind("/");
+    if (subnet_pos < 0) {
+      throw MOADNSException("Asked to decode '"+zone+"' as an APL record, but missing subnet mask");
+    }
+
+    // Parse IPv6 string into ComboAddress
+    ipstr = record.substr(2, subnet_pos - 2);
+    subnetstr = record.substr(subnet_pos + 1, record.length() - subnet_pos - 1);
+    ca = makeComboAddress(ipstr);
+
+    // Section 4.2 of RFC 3123 (don't send trailing "0" bytes)
+    // Remove trailing "0" bytes from packet and reduce length
+    bytes = 16; // Start by assuming we'll send 16 bytes
+    for (int i=15; i>=0; i--) {
+      if (ca.sin6.sin6_addr.s6_addr[i] == 0) {
+        // trailing 0 byte, calculate length
+        bytes--;
+      } else {
+        // Found non-0 byte, stop trimming
+        break;
+      }
+    }
+    bzero(ret->d_ip6, sizeof(ret->d_ip6));
+    for (int i=0; i<bytes; i++) {
+      ret->d_ip6[i] = ca.sin6.sin6_addr.s6_addr[i];
+    }
+    ret->d_afdlength = bytes;
+
+    // Prefix length
+    prefix = 0;
+    if (sscanf(subnetstr.c_str(), "%u", &prefix) == 1) {
+      ret->d_prefix=prefix;
+    } else
+          throw MOADNSException("Asked to decode '"+zone+"' as an IPv6 APL record, but found invalid prefix string "+subnetstr);
+    if (prefix > 32)
+      throw MOADNSException("Asked to decode '"+zone+"' as an IPv6 APL record, but found invalid prefix value "+subnetstr);
+   } else {
+      throw MOADNSException("Asked to encode '"+zone+"' as an IPv6 APL record but got unknown Address Family");
+  }
+
+  return ret;
+}
+
+// DNSRecord to Packet conversion
+void APLRecordContent::toPacket(DNSPacketWriter& pw) {
+
+    pw.xfr16BitInt(d_family);
+    pw.xfr8BitInt(d_prefix);
+    pw.xfr8BitInt((d_n << 7) + d_afdlength);
+    if (d_family == APL_FAMILY_IPV4) {
+      for (int i=0; i<d_afdlength; i++) {
+        pw.xfr8BitInt(d_ip4[i]);
+      }
+    } else if (d_family == APL_FAMILY_IPV6) {
+      for (int i=0; i<d_afdlength; i++) {
+        pw.xfr8BitInt(d_ip6[i]);
+      }
+    }
+}
+
+// Decode record into string
+string APLRecordContent::getZoneRepresentation(bool noDot) const {
+  string s_n, s_family, s_ip;
```

Comment:
this whole thing could be just Netmask#toString

Reply:
Switched to Netmask class.